File: suites/01_algorithms/comparisons/binary_search.cpp
#include <cstdio>

static long long binary_search(long long limit, long long target) {
    long long low = 0;
    long long high = limit;
    while (low < high) {
        long long mid = low + ((high - low) / 2);
        long long value = (mid * 3) + 7;
        if (value == target) {
            return mid;
        }
        if (value < target) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    return -1;
}

int main() {
    printf("%lld\n", binary_search(10000000, 2999998));
    return 0;
}

File: suites/01_algorithms/comparisons/binary_search.rs
fn binary_search(limit: i64, target: i64) -> i64 {
    let mut low: i64 = 0;
    let mut high: i64 = limit;
    while low < high {
        let mid = low + ((high - low) / 2);
        let value = (mid * 3) + 7;
        if value == target {
            return mid;
        }
        if value < target {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    -1
}

fn main() {
    println!("{}", binary_search(10000000, 2999998));
}
